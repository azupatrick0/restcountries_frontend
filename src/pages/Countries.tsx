/* eslint-disable react-hooks/exhaustive-deps */
import React, { FC, useState, useEffect } from 'react';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import NavBar from 'components/NavBar';
import CountryCard from 'components/Card/CountryCard';
import Loader from 'components/Loader';
import fetchAllCountries from 'actions/fetchAllCountries';
import searchCountries from 'actions/searchCountries';

export interface iState {
	countries: any;
	loading: boolean;
	error: string;
}

const Countries: FC = () => {
	const [ searchValue, setSearchValue ] = useState('');
	const state = useSelector((state: iState) => state, shallowEqual);
	const dispatch = useDispatch();

	useEffect(() => {
		dispatch(fetchAllCountries());
	}, []);

	const handleSearchTermChange = (e: { target: { value: React.SetStateAction<string> } }) => {
		setSearchValue(e.target.value);
	};

	const handleSearch = () => {
    dispatch(searchCountries(searchValue));
    setSearchValue('');
	};

	return (
		<div className="countries">
			<NavBar
				searchValue={searchValue}
				handleSearch={handleSearch}
				handleSearchTermChange={handleSearchTermChange}
			/>
			<div className="cards__layout">
				{state.countries.allCountries && state.countries.loading ? (
					<Loader />
				) : state.countries.allCountries && state.countries.allCountries.length ? (
					state.countries.allCountries.map((country: any): any => (
						<CountryCard
							name={country.name}
							flag={country.flag}
							region={country.region}
							key={country.name}
						/>
					))
				) : (
					'No countries found'
				)}
			</div>
		</div>
	);
};

export default Countries;
