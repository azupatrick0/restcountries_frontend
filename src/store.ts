import { createStore, applyMiddleware, combineReducers } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import fetchAllCountries from 'reducers/fetchAllCountries';

const rootReducer = combineReducers({
  countries: fetchAllCountries,
});

const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk)));

export default store;