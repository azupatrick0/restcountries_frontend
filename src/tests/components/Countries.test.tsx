import React from 'react';
import { mount } from 'enzyme';
import Countries from 'pages/Countries';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from 'store';

it('should render countries page', () => {
	const wrapper = mount(
		<MemoryRouter>
			<Provider store={store}>
				<Countries />
			</Provider>
		</MemoryRouter>
	);

  expect(wrapper.find('.countries').text()).toContain('SEARCH');
  expect(wrapper).toMatchSnapshot();
});

it('should set the search value on change event', () => {
	const wrapper = mount(
		<MemoryRouter>
			<Provider store={store}>
				<Countries />
			</Provider>
		</MemoryRouter>
	);

	wrapper.find('.search__input').simulate('change', {
    target: {
      value: 'Malta'
    },
  });
  
	expect(wrapper.find('.search__input').prop('value')).toEqual('Malta');
});

it('should search for countries when search button is clicked', () => {
  const wrapper = mount(
		<MemoryRouter>
			<Provider store={store}>
				<Countries />
			</Provider>
		</MemoryRouter>
  );

  wrapper.find('.search__input').simulate('change', {
    target: {
      value: 'Malta'
    },
  });
  
  wrapper.find('.search').find('button').simulate('click');
  
  expect(wrapper.find('.search__input').text()).toEqual('');
});
