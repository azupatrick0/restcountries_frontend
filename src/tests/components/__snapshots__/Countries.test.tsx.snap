// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render countries page 1`] = `ReactWrapper {}`;
